#   **Homework 8 and Summary Report**
##   Data Fluency | Fall 2024
##   December 2, 2024
##   Student: Joseph Juarez

### **This assignment will utilize the Bayes classifier provided by Python to perform a simple classification task.**


---

### **Step 1 (5 Points)**
#### Prepare a breast cancer data set that is provided by Python datasets package. 
#### You may split this dataset into training set and testing set according to your own splitting ratio.
#### The following picture is a code example how to load the cancer dataset into your Python file.
#### You can print(data.data) and print(data.target) to preview the features and labels.
#### You can also print(data) to preview the whole dataset.

#### **Note: Make sure to install this via terminal:**
    pip install scikit-learn

In [ ]:
from sklearn.datasets import load_breast_cancer
data = load_breast_cancer()

print(data)
print("\n")
print(data.data[[10, 50, 85]])
print("\n")
print(data.target[[10, 50, 85]])
print("\n")
print(data.target_names)

{'data': array([[1.799e+01, 1.038e+01, 1.228e+02, ..., 2.654e-01, 4.601e-01,
        1.189e-01],
       [2.057e+01, 1.777e+01, 1.329e+02, ..., 1.860e-01, 2.750e-01,
        8.902e-02],
       [1.969e+01, 2.125e+01, 1.300e+02, ..., 2.430e-01, 3.613e-01,
        8.758e-02],
       ...,
       [1.660e+01, 2.808e+01, 1.083e+02, ..., 1.418e-01, 2.218e-01,
        7.820e-02],
       [2.060e+01, 2.933e+01, 1.401e+02, ..., 2.650e-01, 4.087e-01,
        1.240e-01],
       [7.760e+00, 2.454e+01, 4.792e+01, ..., 0.000e+00, 2.871e-01,
        7.039e-02]]), 'target': array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 0,
       1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1,
       1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0,
 

---

### **Step 2 (20 Points)**
#### Use the GaussianNB that we learned in the class to train Naïve Bayes classifier with the training set and then test this model with your testing set. 
#### (Hint: use the similar code example in the class notes)

In [3]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Splitting the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(data.data, data.target, test_size=0.3, random_state=42)

# Initializing and training the Gaussian Naive Bayes model
model = GaussianNB()
model.fit(X_train, y_train)

# Predicting on the test set
y_pred = model.predict(X_test)

# Calculating accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2%}")


Accuracy: 94.15%


---

### **Step 3 (10 Points)**
#### Please also use classification_report library (same as we discussed in the class notes) to draw a short report in Python.

In [4]:
from sklearn.metrics import classification_report

# Generating the classification report
report = classification_report(y_test, y_pred, target_names=data.target_names)
print("Classification Report:\n", report)


Classification Report:
               precision    recall  f1-score   support

   malignant       0.93      0.90      0.92        63
      benign       0.95      0.96      0.95       108

    accuracy                           0.94       171
   macro avg       0.94      0.93      0.94       171
weighted avg       0.94      0.94      0.94       171



---

## Step 4 (15 Points)
###   **Summary Report**
#### 1. What kind of data object that is applied to save the breast cancer dataset?
    Object Type: <class 'sklearn.utils._bunch.Bunch'>
#### 2. What is the data size (number of recodes, you can use proper Python function to find it or by some other ways)?
    Number of records: 569
#### 3. What is the feature size? (Number of attributes in each data record)
    Feature Size: 30
#### 4. What are labels in this dataset? (list the labels name)
    ['malignant' 'benign']
#### 5. What are the actual feature names used in this dataset? (Hint: you need to use some operation of dictionary to find all feature name in this dataset or by some other ways)
    ['mean radius' 'mean texture' 'mean perimeter' 'mean area'
    'mean smoothness' 'mean compactness' 'mean concavity'
    'mean concave points' 'mean symmetry' 'mean fractal dimension'
    'radius error' 'texture error' 'perimeter error' 'area error'
    'smoothness error' 'compactness error' 'concavity error'
    'concave points error' 'symmetry error' 'fractal dimension error'
    'worst radius' 'worst texture' 'worst perimeter' 'worst area'
    'worst smoothness' 'worst compactness' 'worst concavity'
    'worst concave points' 'worst symmetry' 'worst fractal dimension']

#### 6. Copy the Python generated result report in this .doc file.
    Classification Report:
    
             precision    recall  f1-score   support

    malignant     0.93      0.90      0.92        63
    benign        0.95      0.96      0.95       108

    accuracy                         0.94       171
    macro avg     0.94      0.93      0.94       171
    weighted avg  0.94      0.94      0.94       171


In [27]:
# Answer to 1
print(type(data))

<class 'sklearn.utils._bunch.Bunch'>


In [28]:
# Answer to 2
print(len(data.data))

569


In [30]:
# Answer to 3
len(data.feature_names)

30

In [31]:
# Answer to 4
print(data.target_names)

['malignant' 'benign']


In [32]:
# Answeer to 5
print(data.feature_names)

['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']
